src/executor/evaluate/evaluated.rs

```diff
@@ -182,126 +146,145 @@ impl TryInto<Value> for Evaluated<'_> {
     }
 }
 
-impl<'a> Evaluated<'a> {
-    pub fn add(&self, other: &Evaluated<'a>) -> Result<Evaluated<'a>> {
-        use Evaluated::*;
-
-        let unreachable = || Err(EvaluateError::UnreachableEvaluatedArithmetic.into());
-
-        let add_literal = |l, other: &Evaluated<'a>| match other {
-            LiteralRef(r) => literal_add(l, r).map(Evaluated::Literal),
-            Literal(r) => literal_add(l, &r).map(Evaluated::Literal),
-            ValueRef(r) => r.add(&r.clone_by(l)?).map(Evaluated::Value),
-            Value(r) => r.add(&r.clone_by(l)?).map(Evaluated::Value),
-            StringRef(_) => unreachable(),
-        };
-
-        let add_value = |l: &data::Value, other: &Evaluated<'a>| match other {
-            LiteralRef(r) => l.add(&l.clone_by(r)?).map(Evaluated::Value),
-            Literal(r) => l.add(&l.clone_by(&r)?).map(Evaluated::Value),
-            ValueRef(r) => l.add(r).map(Evaluated::Value),
-            Value(r) => l.add(&r).map(Evaluated::Value),
-            StringRef(_) => unreachable(),
-        };
-
-        match self {
-            LiteralRef(l) => add_literal(l, other),
-            Literal(l) => add_literal(&l, other),
-            ValueRef(l) => add_value(l, other),
-            Value(l) => add_value(&l, other),
-            StringRef(_) => unreachable(),
+macro_rules! binary_op {
+    ($name:ident, $op:tt) => {
+        pub fn $name(&self, other: &Evaluated<'a>) -> Result<Evaluated<'a>> {
+            let literal_binary_op = |l: &&AstValue, r: &&AstValue| match (l, r) {
+                (AstValue::Number(a), AstValue::Number(r)) => match (a.parse::<i64>(), r.parse::<i64>()) {
+                    (Ok(a), Ok(r)) => Ok(AstValue::Number((a $op r).to_string())),
+                    (Ok(a), _) => match r.parse::<f64>() {
+                        Ok(r) => Ok(AstValue::Number(((a as f64) $op r).to_string())),
+                        _ => Err(EvaluateError::UnreachableLiteralArithmetic.into()),
+                    },
+                    (_, Ok(r)) => match a.parse::<f64>() {
+                        Ok(a) => Ok(AstValue::Number((a $op (r as f64)).to_string())),
+                        _ => Err(EvaluateError::UnreachableLiteralArithmetic.into()),
+                    },
+                    (_, _) => match (a.parse::<f64>(), r.parse::<f64>()) {
+                        (Ok(a), Ok(r)) => Ok(AstValue::Number((a $op r).to_string())),
+                        _ => Err(EvaluateError::UnreachableLiteralArithmetic.into()),
+                    },
```

Comment:
You might want to change `a` to `l` :)

Reply:
Ah yes, thought I saw that then I couldn't find it again! Haha